Stats: Refactor navigation arrows into a shared component.

a8ck3n opened the pull request:
#### Proposed Changes

Extracts the navigation arrows into a new component.

- Consolidates styles
- Fixes bug in annual insights nav that triggered multiple GA events in response to keyboard events
- Fixes bug in stats chart nav that allowed you to page into the future via keyboard

Looks that same as it did prior to refactor. The navigation in question…

**Traffic → Chart nav:**

![NavChart](https://user-images.githubusercontent.com/40267301/211002250-7c5ff7d2-92c8-4218-b93f-f324d745f944.png)

**Insights → Year in review nav:**

![InsightsNav](https://user-images.githubusercontent.com/40267301/211002338-2331712a-42ce-4daa-a048-f2b1cd0aa3cf.png)

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Spin up the live branch.
* Visit the Traffic page and confirm the arrows look and behave as they did previously. You can compare by viewing the same page in wordpress.com
* Visit the Insights page and confirm the arrows look and behave correctly there too.
* In both cases, it should be possible to tab to the arrows and trigger them via the keyboard (in Chrome, Firefox and Safari require special setup first).
* On the Traffic page, it should not be possible to arrow past the current time period (into the future).
* On the Insights page, it should not be possible to arrow into 2024.

**Note:** This component hasn't been moved into Storybook. If we want to do that, we can tackle it in a follow-up PR.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #71677.

a8ck3n committed:
Extract navigation arrows into new component.

a8ck3n committed:
Localize event handling.

This highlights the fact that we are triggering multiple events per key press instead of just one.

a8ck3n committed:
Remove key tracking.

Modern browsers should handle mapping of Return and Space when configured correctly.

a8ck3n committed:
Separate event callbacks.

a8ck3n committed:
Minor syntax tweak.

a8ck3n committed:
Update traffic chart to use new component.

a8ck3n committed:
Consolidate styles for new component.

a8ck3n committed:
Remove empty function call.

a8ck3n committed:
Move instance-specific styling to caller.

a8ck3n committed:
Remove debug cruft.

a8ck3n committed:
Put new arrow component behind showArrows prop.

a8ck3n committed:
Remove inline implementation from chart nav.

a8ck3n committed:
Rename arrow func for consistency.

a8ck3n committed:
Remove styles that are no longer in use.

a8ck3n committed:
Remove remaining debug logs.

dognose24 commented on client/my-sites/stats/navigation-arrows/style.scss:
```diff
@@ -0,0 +1,32 @@
+@import "@automattic/typography/styles/variables";
+
+.arrow-navigation {
```
Could we use `.navigation-arrows` or `.stats-navigation-arrows` to align with the component name?

a8ck3n replied:
Will do! I'm pretty sure you've shared similar feedback in the past so I'll try and pay closer attention to this going forward.

wilmersondasilva reviewed (approved):
Looks good and worked as described!

dognose24 commented on client/my-sites/stats/navigation-arrows/style.scss:
```diff
@@ -0,0 +1,32 @@
+	}
+
+	@media ( min-width: 661px ) {
+		padding-right: 0;
```
This `padding-right` was trying to eliminate the [gap](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/stats/stats-period-navigation/style.scss#L6) of `.stats-period-navigation`, which seems not to be necessary for this PR.

dognose24 replied:
However, the padding to the right edge on `mobile (viewport width < 660px)` would disappear in this PR.

| Before | This PR |
| --- | --- |
| <img width="615" alt="截圖 2023-01-06 下午11 17 29" src="https://user-images.githubusercontent.com/6869813/211042979-093e5020-b07f-43a9-9549-6ce4f63f9e5f.png"> | <img width="611" alt="截圖 2023-01-06 下午11 16 34" src="https://user-images.githubusercontent.com/6869813/211043041-f5eff4cc-4cde-4729-9d95-5dfd89498637.png"> |

a8ck3n replied:
Thanks Dognose! I'll fix these up today.

wilmersondasilva commented:
One question, are the **Store** and **Ads** tabs affected by this change? Does their code need some change?

jsnmoon commented on client/my-sites/stats/navigation-arrows/index.jsx:
```diff
@@ -0,0 +1,42 @@
+	disableNextArrow,
+	disablePreviousArrow,
+	onArrowNext,
+	onArrowPrevious,
```
```suggestion
	onClickNext,
	onClickPrevious,
```

jsnmoon commented on client/my-sites/stats/navigation-arrows/index.jsx:
```diff
@@ -0,0 +1,42 @@
+				className={ classNames( 'arrow-navigation__previous', {
+					'is-disabled': disablePreviousArrow,
+				} ) }
+				onClick={ disablePreviousArrow ? undefined : () => handleClickPrevious() }
```
```suggestion
				onClick={ disablePreviousArrow ? undefined : onClickPrevious }
```

Let's bind the entire function to the event so that component consumers can directly access all event callback parameters. As a bonus, this will help us get rid of the intermediate callback functions`handleClickPrevious` and `handleClickNext` 🙂

jsnmoon commented on client/my-sites/stats/navigation-arrows/index.jsx:
```diff
@@ -0,0 +1,42 @@
+				className={ classNames( 'arrow-navigation__next', {
+					'is-disabled': disableNextArrow,
+				} ) }
+				onClick={ disableNextArrow ? undefined : () => handleClickNext() }
```
```suggestion
				onClick={ disableNextArrow ? undefined : onClickNext }
```

jsnmoon commented on client/my-sites/stats/navigation-arrows/index.jsx:
```diff
@@ -0,0 +1,42 @@
+				className={ classNames( 'arrow-navigation__previous', {
+					'is-disabled': disablePreviousArrow,
+				} ) }
+				onClick={ disablePreviousArrow ? undefined : () => handleClickPrevious() }
```
In a follow-up PR, we'll also want to handle `onKeyDown` for `enter` key presses. (I believe this was part of the YearNavigation component in the annual highlights section.)

a8ck3n replied:
My understanding is this is no longer needed for modern browsers as they will properly map the `space` and `return` keys to a button click. I tested this in Chrome and it worked as expected. I'll give it another look and include Firefox and Safari in my testing.

a8ck3n replied:
I've opened [a follow up Issue](https://github.com/Automattic/wp-calypso/issues/71836) for this.

jsnmoon replied:
Ah, I had no idea! Thanks for creating a follow up issue.

grzegorz-cp reviewed (commented):
It looks good. Thanks for consolidating the code @a8ck3n. The only thing I noticed was what @dognose24 pointed out already - below 660px on the Insights page arrows don't have enough padding.

a8ck3n committed:
Update class names to reflect component name.

Instances of `arrow-navigation` → `stats-navigation-arrows`.

a8ck3n committed:
Remove unused style.

a8ck3n committed:
Fix right-hand margin on smaller screens.

dognose24 commented on client/my-sites/stats/annual-highlights-section/style.scss:
```diff
@@ -1,15 +1,11 @@
 
-	@media ( min-width: 661px ) {
-		padding-right: 0;
+		@media ( max-width: 661px ) {
```
```suggestion
		@media ( max-width: $custom-mobile-breakpoint ) {
```

This breakpoint could be related to the [highlight-cards variable](https://github.com/Automattic/wp-calypso/blob/trunk/packages/components/src/highlight-cards/variables.scss#L6).

a8ck3n replied:
Thanks for the heads up. I'll get this updated.

dognose24 reviewed (approved):
LGTM! I am only concerned about reusing the breakpoint variable, which could not be a blocker. 🙂

kangzj commented on client/my-sites/stats/navigation-arrows/index.jsx:
```diff
@@ -0,0 +1,42 @@
+	onArrowNext,
+	onArrowPrevious,
+} ) {
+	const handleClickNext = () => {
+		onArrowNext();
+	};
+	const handleClickPrevious = () => {
+		onArrowPrevious();
+	};
```
*1

kangzj commented on client/my-sites/stats/navigation-arrows/index.jsx:
```diff
@@ -0,0 +1,42 @@
+				className={ classNames( 'stats-navigation-arrows__previous', {
+					'is-disabled': disablePreviousArrow,
+				} ) }
+				onClick={ disablePreviousArrow ? undefined : () => handleClickPrevious() }
```
I'm just wondering why the functions are wrapper by arrow functions a second time here (firstly in *1), is it necessary? IMHO, the first wrapping (*1) is not necessary as well or am I missing something?

a8ck3n replied:
You're right! I like to have a local handler for the purpose of logging (during development) or pre-processing. I should have caught that all the logic was removed and just used the callbacks from props. I'll get this fixed up!

kangzj commented on client/my-sites/stats/navigation-arrows/index.jsx:
```diff
@@ -0,0 +1,42 @@
+				className={ classNames( 'stats-navigation-arrows__next', {
+					'is-disabled': disableNextArrow,
+				} ) }
+				onClick={ disableNextArrow ? undefined : () => handleClickNext() }
```
Ditto as the previous comment.

kangzj reviewed (approved):
It works well for me. Left some questions, no blockers. Good one 👍

jsnmoon reviewed (commented):
I left some non-blocking suggestions below, specifically regarding property names and how we're binding event handlers for this component.

a8ck3n commented:
> One question, are the **Store** and **Ads** tabs affected by this change? Does their code need some change?

Sorry @wilmersondasilva, I missed replying to this. No codes changes are needed in the **Store** or **Ads** tabs, at least not yet. If there are navigation arrows on those tabs, I'll spin up a new issue to address those.

a8ck3n committed:
Use CSS variable for mobile breakpoint.

a8ck3n committed:
Simplify onClick handlers.

No need for the local hanlder any more. Just use the callbacks provided via props.
